[merged] Host object now house a remote ssh username.

ashcrow opened the pull request:
This will allow OS's and cloud providers which block the use of root
ssh. If no ```remote_user``` is passed on host object creation root is used.
This change also introduces ```authorized_keys_file``` and ```remote_user``` to the
```cloud-init``` integration.

ashcrow commented:
@mbarnes if you could give this an early review I'd appreciate it. It's not ready to merge just yet but I *think* the majority of it is done and ready to go.

mbarnes commented on contrib/cloud-init/commissaire.txt:
```diff
@@ -1,7 +1,9 @@
+  "password": "12345",                                 # Optional
+  "cluster": "mycluster",                              # Optional
+  "remote_user": "root",                               # Optional
+  "root_ssh_key_path": "/root/.ssh/id_rsa",            # Optional
```
Maybe rename this to `ssh_key_path` or `priv_ssh_key_path` since it's not necessarily root now?

Future feature: Would be nice to allow `$HOME` in the config file paths, so they're more orthogonal with `remote_user`.

mbarnes commented on doc/endpoints.rst:
```diff
@@ -307,6 +307,7 @@ Creates a new host record.
 
    {
        "ssh_priv_key": string // base64 encoded ssh private key
+       "remote_user": string  // Optional name of ssh user to use
```
Since this has a non-empty default, I'd clarify as:
`// Name of ssh user to use (default=root)`

mbarnes commented on src/commissaire/transport/ansibleapi.py:
```diff
@@ -139,12 +140,23 @@ def _run(self, ips, key_file, play_file,
-            become=None, become_method=None, become_user=None,
+            become=become['become'], become_method=None,
+            become_user=become['become_user'],
             verbosity=None, check=False)
```
May be simpler to assemble `options` as a dictionary, then futher down in the `TaskQueueManager` call do `options=self.Options(**options)`.

mbarnes replied:
Also, `become_method` should be `"su"` or `"sudo"`.  That appears to be causing @cooktheryan's exception below.

mbarnes commented:
So far so good!  :+1:

cooktheryan commented:
It looks like we need to restorecon -R -v /home/remote_user/.ssh I just can't code in python to save my life

cooktheryan commented:
Someone please clean this up but it get us farther 
`    if not os.path.isfile(keyfile):
        try:
            subprocess.check_call(
                ['/usr/bin/ssh-keygen', '-q', '-N', '',
                 '-t', 'rsa', '-f', keyfile])
            subprocess.check_call(
                ['/sbin/restorecon', '-R', '-v',
                '/home/commissaire-user/.ssh'])
            subprocess.check_call(
                ['/bin/chown', '-R', remote_user,
                '/home/commissaire-user'])
        except FileNotFoundError:
            print('Missing /usr/bin/ssh-keygen', file=sys.stderr)
            raise
        except subprocess.CalledProcessError as ex:
            print(str(ex), file=sys.stderr)
            raise`

We now can connect the host but are hitting the following exception 
https://gist.github.com/cooktheryan/2f08d8c548b0acb49e4c81a98ad91b14

mbarnes commented on contrib/cloud-init/part-handler.py:
```diff
@@ -115,6 +115,13 @@ def handle_part(data, ctype, filename, payload):
 
+    # Verify the authorized_keys file exists
+    if not os.path.isfile(authorized_keys):
+        authorized_keys_path, _ = authorized_keys.rsplit('/', 1)
```
`os.path.dirname(authorized_keys)` would be better here, no?

mbarnes commented on contrib/cloud-init/part-handler.py:
```diff
@@ -115,6 +115,13 @@ def handle_part(data, ctype, filename, payload):
+        authorized_keys_path, _ = authorized_keys.rsplit('/', 1)
+        os.makedirs(authorized_keys_path)
+        with open(authorized_keys, 'w') as _:
+            pass
```
I can see creating the directory, but a little further down `authorized_keys` is opened with `os.O_CREAT` and mode 0600.  This `open` call looks redundant.

mbarnes commented:
(copying my comment on Ryan's branch here)

Could do the file ownership recursively in Python with `os.walk` and `os.chown`, but this might be just as easy.

For the user's home path: `os.path.expanduser('~' + remote_user)`
Python will look up `remote_user` in the password file.

ashcrow commented:
I'll clean it up and amend it to the WIP for your testing.

ashcrow commented:
@cooktheryan  I implemented your gist into the part-handler. Fire away.

ashcrow commented:
I got confirmation from @cooktheryan that things are working as he would expect them to. I've squashed the fixups and fixed some history.

@mbarnes :arrow_up:

cooktheryan commented:
@rh-atomic-bot +try

cooktheryan commented:
@rh-atomic-bot +r

mbarnes commented:
@ashcrow nice!
